### PR TITLE
Implement Shanghai

### PR DESCRIPTION
**Pull request by gurukamath**
### What was wrong?
Shanghai Hardfork not implemented

### How was it fixed?
Implemented EIP-4895, EIP-3651, EIP-3860 and EIP-3855

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by gurukamath** [comment]
Currently in draft because the tests for some EIPs are in `Merge+EIP` phase rather than `Shanghai`

File: src/ethereum/shanghai/spec.py [hunk 1]
```diff
@@ -573,7 +574,15 @@ def validate_transaction(tx: Transaction) -> bool:
     verified : `bool`
         True if the transaction can be executed, or False otherwise.
     """
-    return calculate_intrinsic_cost(tx) <= tx.gas and tx.nonce < 2**64 - 1
+    valid_transaction = True
```
Comment (petertdavies) [inline src/ethereum/shanghai/spec.py:577]:
```python
if ...:
    return False
elif ...:
    return False

return True
```
would be clearer and more idiomatic.

**Review by petertdavies — CHANGES_REQUESTED** [review]
I found a couple of nits.

File: src/ethereum/shanghai/spec.py (hunk 1 above)
Comment (gurukamath) [reply src/ethereum/shanghai/spec.py]: Updated

File: src/ethereum/shanghai/state.py [hunk 2]
```diff
@@ -453,6 +453,20 @@ def increase_recipient_balance(recipient: Account) -> None:
...
+    wd: Withdrawal,
+) -> None:
+    """
+    Increase the balance of the withdrawing account.
+    """
+
+    def increase_recipient_balance(recipient: Account) -> None:
+        recipient.balance += wd.amount * 10**9
+
+    modify_state(state, wd.address, increase_recipient_balance)
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/shanghai/state.py:467]: The `create_ether` function was previously used for issuing mining rewards and is no longer relevant post merge. Having something more closely related to the context of withdrawals makes more sense to me in this case. I am removing the `create_ether` function from Shanghai.

Will also create a new issue to remove the function from Paris

**Review by petertdavies — APPROVED** [review]

**Pull request merged**